### PR TITLE
Adds basic sequence diagrams behind a layout flag

### PR DESCRIPTION
**Pull request by ejulio-ts**
Adds `d2sequence` layout for sequence diagrams with basic functionality.
The layout is behind a flag for now as it will need changes in `D2` for full support. So, to merge these changes in small PRs, lets have it behind a flag that also allows us to test it locally without much friction.

Assumptions:
- edges are kept in order as they are defined
- object order is identified as they appear in edge order
- the lifeline edge follows the shape stroke style, except for `stroke-dash` that is constant

No changelog as this is not ready yet. Will add the line once we got it properly integrated.

here is an example diagram

```
alice: "Alice\nline\nbreaker" {
    shape: person
    style.stroke: red
}
bob: "Bob" {
    shape: person
    style.stroke-width: 5
}
db: {
    shape: cylinder
}
queue: {
    shape: queue
}
service: "an\nodd\nservice\nwith\na\nname\nin\nmultiple lines"

alice -> bob: "Authentication Request"
bob -> service: "make request"
service -> db: "validate credentials"
db -> service: {
    style.stroke-dash: 4
}
service -> bob: {
    style.stroke-dash: 4
}
bob -> alice: "Authentication Response"
alice -> bob: "Another authentication Request"
bob -> queue: "do it later"
queue -> bob: "stored" {
    style.stroke-dash: 3
    style.stroke-width: 5
    style.stroke: green
}
bob -> alice: "Another authentication Response"
```

_Users_juliobatista_Projects_terrastruct_d2_temp svg (1)

Updates #99

### CODE REVIEW

File: d2layouts/d2sequence/layout.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	"oss.terrastruct.com/d2/lib/geo"
+	"oss.terrastruct.com/d2/lib/go2"
+	"oss.terrastruct.com/d2/lib/label"
+)
+
+func Layout(ctx context.Context, g *d2graph.Graph) (err error) {
+	pad := 50. // 2 * 25
+	edgeYStep := 100.
+	objectXStep := 200.
+	maxObjectHeight := 0.
```
Comment (alixander) [inline d2layouts/d2sequence/layout.go:18]: we should put all constants in a file. likely we'll want this to be configurable by power users

File: d2layouts/d2sequence/layout.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+	var objectsInOrder []*d2graph.Object
+	seen := make(map[*d2graph.Object]struct{})
```
Comment (alixander) [inline d2layouts/d2sequence/layout.go:21]: u shouldn't need to do this. the graph's objects are already in order by when they first appear

File: d2layouts/d2sequence/layout.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	addLifelineEdges(g, objectsInOrder, edgeYStep)
+
+	return nil
+}
+
+// placeObjects places objects side by side
+func placeObjects(objectsInOrder []*d2graph.Object, maxHeight, xStep float64) {
+	x := 0.
+	for _, obj := range objectsInOrder {
+		yDiff := maxHeight - obj.Height
```
Comment (alixander) [inline d2layouts/d2sequence/layout.go:50]: we should bottom-align these. makes a cleaner appearance

File: d2layouts/d2sequence/layout.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		start := edge.Src.Center()
+		start.Y = edgeY
+		end := edge.Dst.Center()
+		end.Y = edgeY
+		edge.Route = []*geo.Point{start, end}
+		edgeY += yStep
+
+		if edge.Attributes.Label.Value != "" {
+			// TODO: consider label right-to-left
+			edge.LabelPosition = go2.Pointer(string(label.OutsideTopCenter))
```
Comment (alixander) [inline d2layouts/d2sequence/layout.go:70]: we should make the edge label for sequence diagrams have fill, so the actor lines don't go through them. can u add an issue

**Review by alixander — CHANGES_REQUESTED** [review]
let's introduce sequence-diagram-specific terminology, starting with s/object/actor

File: d2layouts/d2sequence/layout.go (hunk 2 above)
Comment (ejulio-ts) [reply d2layouts/d2sequence/layout.go:21]: What if the user defines the objects upfront?
```
a: {}
b: {}
c: {}

c -> a
b -> c
b -> a
```

File: d2layouts/d2sequence/layout.go (hunk 4 above)
Comment (ejulio-ts) [reply d2layouts/d2sequence/layout.go:70]: Do we have support for it?
If not, should I just add support here or think of the larger D2 change to support label fills?

Comment (alixander) [reply d2layouts/d2sequence/layout.go:70]: we don't have support for it. I think it should be broader. In fact, @gavin-ts will take

File: d2layouts/d2sequence/constants.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package d2sequence
+
+// min horizontal pad for actors, or edge labels, to consider the min distance between them
```
Comment (alixander) [inline d2layouts/d2sequence/constants.go:3]: this isn't clear

File: d2layouts/d2sequence/layout.go [hunk 6]
```diff
@@ -12,45 +12,33 @@ import (
 )
 
 func Layout(ctx context.Context, g *d2graph.Graph) (err error) {
-	pad := 50. // 2 * 25
-	edgeYStep := 100.
-	objectXStep := 200.
-	maxObjectHeight := 0.
+	pad := MIN_HORIZONTAL_PAD
```
Comment (alixander) [inline d2layouts/d2sequence/layout.go:15]: u don't need to reassign constants that don't change

**Review by alixander — CHANGES_REQUESTED** [review]

File: cmd/d2/main.go [hunk 7]
```diff
@@ -187,8 +189,13 @@ func compile(ctx context.Context, ms *xmain.State, plugin d2plugin.Plugin, theme
 		return nil, err
 	}
 
+	layout := plugin.Layout
+	// TODO: remove, this is just a feature flag to test sequence diagrams as we work on them
+	if os.Getenv("D2_SEQUENCE") == "1" {
```
Comment (alixander) [inline cmd/d2/main.go:194]: interesting. maybe we should have an EXPERIMENTS flag. I could see users wanting to use this even if the support is early and unstable. I'll make an issue on this.

Comment (alixander) [reply cmd/d2/main.go:194]: https://github.com/terrastruct/d2/issues/241

**Comment by ejulio-ts** [comment]
_Users_juliobatista_Projects_terrastruct_d2_temp svg (2)

**Review by alixander — APPROVED** [review]

**Pull request merged**